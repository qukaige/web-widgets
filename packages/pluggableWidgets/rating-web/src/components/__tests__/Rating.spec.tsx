import { createElement } from "react";
import { mount, shallow } from "enzyme";
import { Rating, RatingProps } from "../Rating";

describe("Rating", () => {
    const defaultProps: RatingProps = {
        className: "",
        animated: true,
        disabled: false,
        emptyIcon: <div className="empty" />,
        fullIcon: <div className="full" />,
        value: 0,
        onChange: undefined,
        maximumValue: 5,
        style: undefined
    };

    it("renders correctly the structure", () => {
        const rating = shallow(<Rating {...defaultProps} />);
        expect(rating).toMatchSnapshot();
    });

    it("renders correctly the structure when disabled", () => {
        const rating = shallow(<Rating {...defaultProps} disabled />);
        expect(rating).toMatchSnapshot();
    });

    it("renders correctly the structure without animations", () => {
        const rating = shallow(<Rating {...defaultProps} animated={false} />);
        expect(rating).toMatchSnapshot();
    });

    it("renders correctly the structure with custom class", () => {
        const rating = shallow(<Rating {...defaultProps} className="my-custom-class" />);
        expect(rating).toMatchSnapshot();
    });

    it("renders the correct amount of items", () => {
        const rating = mount(<Rating {...defaultProps} maximumValue={2} />);
        expect(rating.find(".rating-item")).toHaveLength(2);
    });

    it("renders the correct amount of items when value is superior to maximumValue", () => {
        const rating = mount(<Rating {...defaultProps} maximumValue={2} value={5} />);
        expect(rating.find(".rating-item")).toHaveLength(2);
        expect(rating.find("div.full")).toHaveLength(2);
    });

    describe("with events", () => {
        it("triggers the event with correct value on click", () => {
            const onChange = jest.fn();
            const rating = mount(<Rating {...defaultProps} onChange={onChange} />);
            const options = rating.find(".rating-item");
            options.at(0).simulate("click");
            expect(onChange).toHaveBeenCalled();
            expect(onChange).toHaveBeenCalledWith(1);
        });

        it("cleans the value when clicking twice at same value", () => {
            const onChange = jest.fn();
            const rating = mount(<Rating {...defaultProps} onChange={onChange} />);
            const options = rating.find(".rating-item");
            options.at(1).simulate("click");
            expect(onChange).toHaveBeenCalledWith(2);
            // As the property is being managed from the outer component, we need to force the property to update to the new value
            rating.setProps({ value: 2 });
            options.at(1).simulate("click");
            expect(onChange).toHaveBeenCalledWith(0);
        });

        it("triggers the event with correct value on space key down", () => {
            const onChange = jest.fn();
            const rating = mount(<Rating {...defaultProps} onChange={onChange} />);
            const options = rating.find(".rating-item");
            options.at(0).simulate("keydown", { key: " " });
            expect(onChange).toHaveBeenCalled();
            expect(onChange).toHaveBeenCalledWith(1);
        });

        it("triggers the event with correct value on enter key down", () => {
            const onChange = jest.fn();
            const rating = mount(<Rating {...defaultProps} onChange={onChange} />);
            const options = rating.find(".rating-item");
            options.at(2).simulate("keydown", { key: "Enter" });
            expect(onChange).toHaveBeenCalled();
            expect(onChange).toHaveBeenCalledWith(3);
        });

        it("doesn't trigger any event when disabled", () => {
            const onChange = jest.fn();
            const rating = mount(<Rating {...defaultProps} disabled onChange={onChange} />);
            const options = rating.find(".rating-item");
            options.at(1).simulate("keydown", { key: "Enter" });
            expect(onChange).toHaveBeenCalledTimes(0);
            options.at(2).simulate("keydown", { key: " " });
            expect(onChange).toHaveBeenCalledTimes(0);
            options.at(3).simulate("click");
            expect(onChange).toHaveBeenCalledTimes(0);
        });
    });
});
