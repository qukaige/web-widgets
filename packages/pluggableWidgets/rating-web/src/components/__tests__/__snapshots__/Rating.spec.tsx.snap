// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating renders correctly the structure 1`] = `
<div
  className="widget-rating"
  data-focusindex={0}
  onKeyDown={[Function]}
  role="radiogroup"
>
  <div
    aria-checked={false}
    aria-label="1"
    className="rating-item"
    key="icon_1"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={0}
  >
    <div
      className="empty"
    />
  </div>
  <div
    aria-checked={false}
    aria-label="2"
    className="rating-item"
    key="icon_2"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="empty"
    />
  </div>
  <div
    aria-checked={false}
    aria-label="3"
    className="rating-item"
    key="icon_3"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="empty"
    />
  </div>
  <div
    aria-checked={false}
    aria-label="4"
    className="rating-item"
    key="icon_4"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="empty"
    />
  </div>
  <div
    aria-checked={false}
    aria-label="5"
    className="rating-item"
    key="icon_5"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="empty"
    />
  </div>
</div>
`;

exports[`Rating renders correctly the structure when disabled 1`] = `
<div
  className="widget-rating"
  data-focusindex={0}
  onKeyDown={[Function]}
  role="radiogroup"
>
  <div
    aria-checked={false}
    aria-label="1"
    className="rating-item disabled"
    key="icon_1"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={0}
  >
    <div
      className="empty"
    />
  </div>
  <div
    aria-checked={false}
    aria-label="2"
    className="rating-item disabled"
    key="icon_2"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="empty"
    />
  </div>
  <div
    aria-checked={false}
    aria-label="3"
    className="rating-item disabled"
    key="icon_3"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="empty"
    />
  </div>
  <div
    aria-checked={false}
    aria-label="4"
    className="rating-item disabled"
    key="icon_4"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="empty"
    />
  </div>
  <div
    aria-checked={false}
    aria-label="5"
    className="rating-item disabled"
    key="icon_5"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="empty"
    />
  </div>
</div>
`;

exports[`Rating renders correctly the structure with custom class 1`] = `
<div
  className="widget-rating my-custom-class"
  data-focusindex={0}
  onKeyDown={[Function]}
  role="radiogroup"
>
  <div
    aria-checked={false}
    aria-label="1"
    className="rating-item"
    key="icon_1"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={0}
  >
    <div
      className="empty"
    />
  </div>
  <div
    aria-checked={false}
    aria-label="2"
    className="rating-item"
    key="icon_2"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="empty"
    />
  </div>
  <div
    aria-checked={false}
    aria-label="3"
    className="rating-item"
    key="icon_3"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="empty"
    />
  </div>
  <div
    aria-checked={false}
    aria-label="4"
    className="rating-item"
    key="icon_4"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="empty"
    />
  </div>
  <div
    aria-checked={false}
    aria-label="5"
    className="rating-item"
    key="icon_5"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="empty"
    />
  </div>
</div>
`;

exports[`Rating renders correctly the structure without animations 1`] = `
<div
  className="widget-rating"
  data-focusindex={0}
  onKeyDown={[Function]}
  role="radiogroup"
>
  <div
    aria-checked={false}
    aria-label="1"
    className="rating-item"
    key="icon_1"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={0}
  >
    <div
      className="empty"
    />
  </div>
  <div
    aria-checked={false}
    aria-label="2"
    className="rating-item"
    key="icon_2"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="empty"
    />
  </div>
  <div
    aria-checked={false}
    aria-label="3"
    className="rating-item"
    key="icon_3"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="empty"
    />
  </div>
  <div
    aria-checked={false}
    aria-label="4"
    className="rating-item"
    key="icon_4"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="empty"
    />
  </div>
  <div
    aria-checked={false}
    aria-label="5"
    className="rating-item"
    key="icon_5"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="empty"
    />
  </div>
</div>
`;
